Calvin Passmore

A02107892

ECE 5420

# Lab Amplifiers

## 1 Prelab

i = 1/2 * Kn * Vov^2

Vgs = Vds

i = (Vdd - Vds) / (Rref)

Vds = Vdd - i*Rref = Vgs

i = 1/2 * Kn * (Vdd - i * R - Vgs) ^ 2

i = 1/2 * Kn * [(Vdd^2 - Vdd * R - Vth*Vdd) + (-Vdd * i * R + i^2 * R^2 + Vth * i * R) + (-Vth * Vdd + Vth * i * R + Vthn^2)]

i = 1/2 * Kn[Vdd^2 + i^2 * R^2 + Vth ^ 2 - 2 * Vdd * i * R + 2 * Vth * i * R - 2 * Vth * Vdd]

0 = i^2(1/2 * Kn * R^2) + i(-2 * Vdd * R + 2 * Vth * R - 1) + (Vdd^2 + Vth^2 - 2 * Vth * Vdd)

i = ((2 * Vdd * r - 2 * Vth * R + 1) +- sqrt[(-2 * Vdd * R + 2 * Vth * r - 1)^2 - 2 * K * R^2 * (Vdd ^ 2 + Vth ^ 2 - 2 * Vth * Vdd)]) / (Kn * R)


In [3]:
from math import sqrt
Vdd = 5 # V
Vth = 0.75
Kn = 0.7e-3 # A
R = 10e3 # Ohm

i_refp = ((2 * Vdd * R - 2 * Vth * R + 1) + sqrt((-2 * Vdd * R + 2 * Vth * R - 1)**2 - 2 * Kn * R**2 * (Vdd**2 + Vth**2 - 2 * Vth * Vdd))) / (Kn * R)
i_refn = ((2 * Vdd * R - 2 * Vth * R + 1) - sqrt((-2 * Vdd * R + 2 * Vth * R - 1)**2 - 2 * Kn * R**2 * (Vdd**2 + Vth**2 - 2 * Vth * Vdd))) / (Kn * R)

print(i_refp)
print(i_refn)

24283.874839036227
2.12516096377554


i_ref = 2.125 mA

gm = sqrt(2 * IB * Kn)

ro = 1 / (&lambda; * IB)

## 2

Here is the results table that I generated

```
0       -5.00000e+00    1.291116e-12    8.008670e-02    -2.19288e+01    2.257040e+00    1.359372e-11    2.440357e-10    5.888259e+12    2.172622e+11
1       -4.50000e+00    4.562770e-08    7.934292e+01    3.799016e+01    3.868502e+03    4.802508e-07    5.566229e-07    1.735504e+08    1.161310e+08
2       -4.00000e+00    6.795952e-05    1.064588e+02    4.054363e+01    6.098432e+05    2.204869e-04    1.200721e-04    8.519460e+05    1.011310e+06
3       -3.50000e+00    2.491448e-04    8.294510e+01    3.837581e+01    1.492871e+06    4.165233e-04    2.200303e-04    3.530310e+05    4.271133e+05
4       -3.00000e+00    4.547100e-04    7.023761e+01    3.693139e+01    2.396698e+06    5.626861e-04    2.916084e-04    2.170066e+05    2.773860e+05
5       -2.50000e+00    6.353315e-04    6.246272e+01    3.591242e+01    3.212763e+06    6.676993e-04    3.411058e-04    1.583717e+05    2.163229e+05
6       -2.00000e+00    7.405122e-04    5.791314e+01    3.525554e+01    3.784932e+06    7.257644e-04    3.665050e-04    1.309852e+05    1.926588e+05
7       -1.50000e+00    7.572604e-04    5.541913e+01    3.487319e+01    4.048810e+06    7.405037e-04    3.703683e-04    1.190680e+05    1.894083e+05
8       -1.00000e+00    7.669354e-04    5.323740e+01    3.452434e+01    4.277513e+06    7.506235e-04    3.725796e-04    1.098886e+05    1.875853e+05
9       -5.00000e-01    7.752391e-04    5.124653e+01    3.419329e+01    4.497689e+06    7.590347e-04    3.744656e-04    1.022302e+05    1.860515e+05
10      0.000000e+00    7.827466e-04    4.940070e+01    3.387466e+01    4.714083e+06    7.662185e-04    3.761616e-04    9.565915e+04    1.846886e+05
11      5.000000e-01    7.896944e-04    4.766377e+01    3.356377e+01    4.930925e+06    7.724354e-04    3.777235e-04    8.988962e+04    1.834469e+05
12      1.000000e+00    7.962122e-04    4.600242e+01    3.325561e+01    5.149585e+06    7.778460e-04    3.791819e-04    8.471235e+04    1.822989e+05
13      1.500000e+00    8.023825e-04    4.438177e+01    3.294409e+01    5.374236e+06    7.825525e-04    3.805568e-04    7.995818e+04    1.812266e+05
14      2.000000e+00    8.082623e-04    4.275979e+01    3.262071e+01    5.610062e+06    7.866128e-04    3.818618e-04    7.547212e+04    1.802178e+05
15      2.500000e+00    8.138936e-04    4.107699e+01    3.227197e+01    5.868826e+06    7.900404e-04    3.831069e-04    7.108454e+04    1.792632e+05
16      3.000000e+00    8.193079e-04    3.923251e+01    3.187292e+01    6.172022e+06    7.927870e-04    3.842998e-04    6.656053e+04    1.783559e+05
17      3.500000e+00    8.245299e-04    3.701496e+01    3.136754e+01    6.564790e+06    7.946772e-04    3.854463e-04    6.146744e+04    1.774904e+05
18      4.000000e+00    8.295776e-04    3.381623e+01    3.058250e+01    7.193957e+06    7.951190e-04    3.865509e-04    5.465628e+04    1.766626e+05
19      4.500000e+00    8.344579e-04    2.586071e+01    2.825281e+01    9.342615e+06    7.898892e-04    3.876154e-04    3.962895e+04    1.758704e+05
```

Here the plots I produced when running plot_tran.m in order (1 - 20)

![](./plot_tran/1.png)

![](./plot_tran/2.png)

![](./plot_tran/3.png)

![](./plot_tran/4.png)

![](./plot_tran/5.png)

![](./plot_tran/6.png)

![](./plot_tran/7.png)

![](./plot_tran/8.png)

![](./plot_tran/9.png)

![](./plot_tran/10.png)

![](./plot_tran/11.png)

![](./plot_tran/12.png)

![](./plot_tran/13.png)

![](./plot_tran/14.png)

![](./plot_tran/15.png)

![](./plot_tran/16.png)

![](./plot_tran/17.png)

![](./plot_tran/18.png)

![](./plot_tran/19.png)

![](./plot_tran/20.png)

Here the plots I generated when running plot_ac.m in order (1 - 20)

![](./plot_ac/1.png)

![](./plot_ac/2.png)

![](./plot_ac/3.png)

![](./plot_ac/4.png)

![](./plot_ac/5.png)

![](./plot_ac/6.png)

![](./plot_ac/7.png)

![](./plot_ac/8.png)

![](./plot_ac/9.png)

![](./plot_ac/10.png)

![](./plot_ac/11.png)

![](./plot_ac/12.png)

![](./plot_ac/13.png)

![](./plot_ac/14.png)

![](./plot_ac/15.png)

![](./plot_ac/16.png)

![](./plot_ac/17.png)

![](./plot_ac/18.png)

![](./plot_ac/19.png)

![](./plot_ac/20.png)

Here are the plots from plot_measurements.m in order of the figure numbers they popped up as

![](./plot_measurements/1_tail_current_vs_vicm.png)

![](./plot_measurements/2_bandwidth_vs_vicm.png)

![](./plot_measurements/3_transconductance_n_type.png)

![](./plot_measurements/4_transconductance_p_type.png)

![](./plot_measurements/5_output_resistance_n_type.png)

![](./plot_measurements/6_output_resistance_p_type.png)

![](./plot_measurements/7_gain_volts.png)

![](./plot_measurements/8_gain_dB.png)

![](./plot_measurements/9_predicted_vs_measured_gain.png)

![](./plot_measurements/10_gain_bandwidth_tradeoff.png)

![](./plot_measurements/11_gain_vs_current.png)

![](./plot_measurements/12_bandwitdh_vs_tail_current.png)

### Conclusions

The accuracy of the predictions were fairly close to the simulated solutions, but there were obvious differences. For instance the transconductance and output resistances were fairly close; however the difference of the predictions vs the simulation gets quite exaggerated when using the resistance and transconductance in calculations like Gain/Current.

The trade of between Bandwidth and Gain is that after about .75 MHz the gain is slowling getting lower. So using this circuit for high frequency systems would be not as useful as using them for low frequency systems.

## 3
Here is the plot for the FFT

![](./ex3_fft.png)

and the calculated SNR is -62.3347

## 4

Changed Vin to 10mV in ex3.sp. Below is the float, and the SNR is -62.3353

![](./ex4_fft.png)

## 5

Below is the fft analysis and the SNR is -62.5993

![](./ex5_fft.png)

## 6

Calculating Av0 and gm

In [9]:
from math import sqrt

Kp = 0.19
Vthp = -0.74
Vdd = 5 # v
Vg = 4
Vgs = Vg - Vdd
Vov = Vgs - Vthp
ID = 1/2 * Kp * Vov**2

gm = sqrt(2 * ID * Kp)
av = -gm * R

print(gm)
print(av)

0.049400000000000006
-494.00000000000006


gm = 0.0494
Av0 = -494

## 7

ex2.sp with opamp.sp: 

```
0       -5.00000e+00    1.291126e-12    1.085331e-02    -3.92888e+01    0.000000e+00    1.359389e-11    2.427459e-10    5.888254e+12    2.183838e+11
1       -4.50000e+00    4.562162e-08    2.239358e-01    -1.29975e+01    0.000000e+00    4.802516e-07    5.563758e-07    1.735504e+08    1.161814e+08
2       -4.00000e+00    6.795952e-05    1.735734e-03    -5.52103e+01    0.000000e+00    2.204869e-04    1.200720e-04    8.519460e+05    1.011311e+06
3       -3.50000e+00    2.491448e-04    5.259033e-04    -6.55819e+01    0.000000e+00    4.165233e-04    2.200303e-04    3.530310e+05    4.271133e+05
4       -3.00000e+00    4.547100e-04    5.926116e-04    -6.45446e+01    0.000000e+00    5.626861e-04    2.916084e-04    2.170066e+05    2.773860e+05
5       -2.50000e+00    6.353315e-04    9.510242e-04    -6.04362e+01    0.000000e+00    6.676993e-04    3.411058e-04    1.583717e+05    2.163229e+05
6       -2.00000e+00    7.405122e-04    2.248149e-04    -7.29635e+01    0.000000e+00    7.257644e-04    3.665050e-04    1.309852e+05    1.926588e+05
7       -1.50000e+00    7.572604e-04    2.131827e-04    -7.34250e+01    0.000000e+00    7.405037e-04    3.703683e-04    1.190680e+05    1.894083e+05
8       -1.00000e+00    7.669354e-04    2.038254e-04    -7.38148e+01    0.000000e+00    7.506235e-04    3.725796e-04    1.098886e+05    1.875853e+05
9       -5.00000e-01    7.752391e-04    1.962282e-04    -7.41448e+01    0.000000e+00    7.590347e-04    3.744656e-04    1.022302e+05    1.860515e+05
10      0.000000e+00    7.827466e-04    1.871757e-04    -7.45550e+01    0.000000e+00    7.662185e-04    3.761616e-04    9.565915e+04    1.846886e+05
11      5.000000e-01    7.896944e-04    1.799763e-04    -7.48957e+01    0.000000e+00    7.724354e-04    3.777235e-04    8.988962e+04    1.834469e+05
12      1.000000e+00    7.962122e-04    1.732769e-04    -7.52252e+01    0.000000e+00    7.778460e-04    3.791819e-04    8.471235e+04    1.822989e+05
13      1.500000e+00    8.023825e-04    1.666835e-04    -7.55622e+01    0.000000e+00    7.825525e-04    3.805568e-04    7.995818e+04    1.812266e+05
14      2.000000e+00    8.082623e-04    1.601143e-04    -7.59114e+01    0.000000e+00    7.866128e-04    3.818618e-04    7.547212e+04    1.802178e+05
15      2.500000e+00    8.138936e-04    1.533663e-04    -7.62854e+01    0.000000e+00    7.900404e-04    3.831069e-04    7.108454e+04    1.792632e+05
16      3.000000e+00    8.193079e-04    1.460529e-04    -7.67098e+01    0.000000e+00    7.927870e-04    3.842998e-04    6.656053e+04    1.783559e+05
17      3.500000e+00    8.245299e-04    1.379676e-04    -7.72045e+01    0.000000e+00    7.946772e-04    3.854463e-04    6.146744e+04    1.774904e+05
18      4.000000e+00    8.295776e-04    1.259132e-04    -7.79986e+01    0.000000e+00    7.951190e-04    3.865509e-04    5.465628e+04    1.766626e+05
19      4.500000e+00    8.344579e-04    9.580314e-05    -8.03724e+01    0.000000e+00    7.898892e-04    3.876154e-04    3.962895e+04    1.758704e+05
```

Transient plots in order of their figure numbers

![](./problem7/plot_tran/1.png)

![](./problem7/plot_tran/2.png)

![](./problem7/plot_tran/3.png)

![](./problem7/plot_tran/4.png)

![](./problem7/plot_tran/5.png)

![](./problem7/plot_tran/6.png)

![](./problem7/plot_tran/7.png)

![](./problem7/plot_tran/8.png)

![](./problem7/plot_tran/9.png)

![](./problem7/plot_tran/10.png)

![](./problem7/plot_tran/11.png)

![](./problem7/plot_tran/12.png)

![](./problem7/plot_tran/13.png)

![](./problem7/plot_tran/14.png)

![](./problem7/plot_tran/15.png)

![](./problem7/plot_tran/16.png)

![](./problem7/plot_tran/17.png)

![](./problem7/plot_tran/18.png)

![](./problem7/plot_tran/19.png)

![](./problem7/plot_tran/20.png)

AC plots in order of their figure number

![](./problem7/plot_ac/1.png)

![](./problem7/plot_ac/2.png)

![](./problem7/plot_ac/3.png)

![](./problem7/plot_ac/4.png)

![](./problem7/plot_ac/5.png)

![](./problem7/plot_ac/6.png)

![](./problem7/plot_ac/7.png)

![](./problem7/plot_ac/8.png)

![](./problem7/plot_ac/9.png)

![](./problem7/plot_ac/10.png)

![](./problem7/plot_ac/11.png)

![](./problem7/plot_ac/12.png)

![](./problem7/plot_ac/13.png)

![](./problem7/plot_ac/14.png)

![](./problem7/plot_ac/15.png)

![](./problem7/plot_ac/16.png)

![](./problem7/plot_ac/17.png)

![](./problem7/plot_ac/18.png)

![](./problem7/plot_ac/19.png)

![](./problem7/plot_ac/20.png)

Plot Measurements:

![](./problem7/plot_measurements/bandwidth_vs_vicm.png)

![](./problem7/plot_measurements/bandwitdh_tail_current.png)

![](./problem7/plot_measurements/gain_dB.png)

![](./problem7/plot_measurements/gain_tradeoff.png)

![](./problem7/plot_measurements/gain_volts.png)

![](./problem7/plot_measurements/output_resistance_ntype.png)

![](./problem7/plot_measurements/output_resistance_p_type.png)

![](./problem7/plot_measurements/predicted_vs_measured_gain.png)

![](./problem7/plot_measurements/tail_current_vs_vicm.png)

![](./problem7/plot_measurements/transconductance_p_type.png)

![](./problem7/plot_measurements/transonductance_n_type.png)

ex3: 

SNR: -62.5993

![](./problem7/ex3.png)

ex4:

SNR: -62.5993

![](./problem7/ex4.png)

## 8

The open loop gain would be Vout/(Vin+Vicm-Vicm) = Vout/Vin

In [11]:
R1 = 1e3
R2 = 10e3

closed_gain = 1 + (R2/R1)
print(closed_gain)

11.0


## 9

running ex2.sp with the added resistors:

```
0       -5.00000e+00    1.291131e-12    5.000013e-01    -6.02058e+00    5.240041e+06    1.359370e-11    2.440341e-10    5.888260e+12    2.172636e+11
1       -4.50000e+00    4.562245e-08    3.599480e+01    3.112480e+01    2.352601e+03    4.802516e-07    5.563800e-07    1.735504e+08    1.161806e+08
2       -4.00000e+00    6.795952e-05    1.258725e+03    6.199861e+01    2.213339e+05    2.204869e-04    1.200721e-04    8.519460e+05    1.011310e+06
3       -3.50000e+00    2.491448e-04    4.463809e+01    3.299411e+01    6.453134e+05    4.165233e-04    2.200303e-04    3.530310e+05    4.271133e+05
4       -3.00000e+00    4.547100e-04    1.081790e+01    2.068286e+01    1.082363e+06    5.626858e-04    2.916081e-04    2.170067e+05    2.773866e+05
5       -2.50000e+00    6.353315e-04    5.716284e+00    1.514228e+01    1.466356e+06    6.676968e-04    3.411038e-04    1.583725e+05    2.163249e+05
6       -2.00000e+00    7.405122e-04    4.059474e+00    1.216939e+01    1.738711e+06    7.257644e-04    3.665050e-04    1.309852e+05    1.926588e+05
7       -1.50000e+00    7.572604e-04    3.468593e+00    1.080307e+01    1.866346e+06    7.405037e-04    3.703683e-04    1.190680e+05    1.894083e+05
8       -1.00000e+00    7.669354e-04    2.992677e+00    9.521197e+00    1.983660e+06    7.506235e-04    3.725796e-04    1.098886e+05    1.875853e+05
9       -5.00000e-01    7.752391e-04    2.576169e+00    8.219487e+00    2.104502e+06    7.590347e-04    3.744656e-04    1.022302e+05    1.860515e+05
10      0.000000e+00    7.827466e-04    2.206006e+00    6.872132e+00    2.231213e+06    7.662185e-04    3.761616e-04    9.565915e+04    1.846886e+05
11      5.000000e-01    7.896944e-04    1.873938e+00    5.455102e+00    2.372117e+06    7.724354e-04    3.777235e-04    8.988962e+04    1.834469e+05
12      1.000000e+00    7.962122e-04    1.574001e+00    3.940101e+00    2.542911e+06    7.778460e-04    3.791819e-04    8.471235e+04    1.822989e+05
13      1.500000e+00    8.023825e-04    1.301645e+00    2.289854e+00    2.772334e+06    7.825525e-04    3.805568e-04    7.995818e+04    1.812266e+05
14      2.000000e+00    8.082623e-04    1.053305e+00    4.510856e-01    3.126418e+06    7.866128e-04    3.818618e-04    7.547212e+04    1.802178e+05
15      2.500000e+00    8.138936e-04    8.261552e-01    -1.65877e+00    3.927282e+06    7.900404e-04    3.831069e-04    7.108454e+04    1.792632e+05
16      3.000000e+00    8.193079e-04    6.179622e-01    -4.18076e+00    9.629904e+06    7.927870e-04    3.842998e-04    6.656053e+04    1.783559e+05
17      3.500000e+00    8.245299e-04    4.270001e-01    -7.39144e+00    0.000000e+00    7.946772e-04    3.854463e-04    6.146744e+04    1.774904e+05
18      4.000000e+00    8.295776e-04    2.519136e-01    -1.19750e+01    0.000000e+00    7.951190e-04    3.865509e-04    5.465628e+04    1.766626e+05
19      4.500000e+00    8.344579e-04    8.870148e-02    -2.10414e+01    0.000000e+00    7.898892e-04    3.876154e-04    3.962895e+04    1.758704e+05
```

Plot Transient in order of their figure number:

![](./problem9/plot_tran/1.png)

![](./problem9/plot_tran/2.png)

![](./problem9/plot_tran/3.png)

![](./problem9/plot_tran/4.png)

![](./problem9/plot_tran/5.png)

![](./problem9/plot_tran/6.png)

![](./problem9/plot_tran/7.png)

![](./problem9/plot_tran/8.png)

![](./problem9/plot_tran/9.png)

![](./problem9/plot_tran/10.png)

![](./problem9/plot_tran/11.png)

![](./problem9/plot_tran/12.png)

![](./problem9/plot_tran/13.png)

![](./problem9/plot_tran/14.png)

![](./problem9/plot_tran/15.png)

![](./problem9/plot_tran/16.png)

![](./problem9/plot_tran/17.png)

![](./problem9/plot_tran/18.png)

![](./problem9/plot_tran/19.png)

![](./problem9/plot_tran/20.png)


Plot AC in order of their figure:

![](./problem9/plot_ac/1.png)

![](./problem9/plot_ac/2.png)

![](./problem9/plot_ac/3.png)

![](./problem9/plot_ac/4.png)

![](./problem9/plot_ac/5.png)

![](./problem9/plot_ac/6.png)

![](./problem9/plot_ac/7.png)

![](./problem9/plot_ac/8.png)

![](./problem9/plot_ac/9.png)

![](./problem9/plot_ac/10.png)

![](./problem9/plot_ac/11.png)

![](./problem9/plot_ac/12.png)

![](./problem9/plot_ac/13.png)

![](./problem9/plot_ac/14.png)

![](./problem9/plot_ac/15.png)

![](./problem9/plot_ac/16.png)

![](./problem9/plot_ac/17.png)

![](./problem9/plot_ac/18.png)

![](./problem9/plot_ac/19.png)

![](./problem9/plot_ac/20.png)

Plot Measurements:

![](./problem9/plot_measurements/bandwidth_vs_tail_current.png)

![](./problem9/plot_measurements/bandwidth_vs_vicm.png)

![](./problem9/plot_measurements/gain.png)

![](./problem9/plot_measurements/gain_bandwidth_tradeoff.png)

![](./problem9/plot_measurements/gain_current.png)

![](./problem9/plot_measurements/gain_dB.png)

![](./problem9/plot_measurements/output_resistance_n_type.png)

![](./problem9/plot_measurements/output_resistance_p_type.png)

![](./problem9/plot_measurements/predicted_vs_measured_gain_relative_error.png)

![](./problem9/plot_measurements/tail_current_vs_vicm.png)

![](./problem9/plot_measurements/transconductance_n_type.png)

![](./problem9/plot_measurements/transconductance_p_type.png)

ex3.sp with the added resistors SNR: -62.5920

![](./problem9/ex3.png)

3x4.sp with the added resistors SNR: -62.5950

![](./problem9/ex4.png)

These results compare relatively well with what was predicted in problem 8, except for non-idealities of the OTA